File: office-addin/src/taskpane/services/oofService.ts
import { getAccessToken } from "./authService";
import { addDays, toISODateStr } from "./subjectHelper";

export interface OofSettings {
  startDate: Date;
  endDate: Date;
  internalMessage: string;
  externalMessage: string;
}

function buildInternalMessage(endDate: Date, signature: string): string {
  const returnDate = addDays(endDate, 1);
  const returnStr = returnDate.toLocaleDateString("en-US", { year: "numeric", month: "long", day: "numeric" });
  const msg = `<p>Thank you for your message. I am currently out of office and will return on ${returnStr}.</p>
<p>I will respond to your email as soon as possible upon my return.</p>
<p>If you need immediate assistance, please contact my colleague.</p>`;
  return signature ? `${msg}${signature}` : msg;
}

function buildExternalMessage(endDate: Date, signature: string): string {
  const returnDate = addDays(endDate, 1);
  const returnStr = returnDate.toLocaleDateString("en-US", { year: "numeric", month: "long", day: "numeric" });
  const returnStrJa = returnDate.toLocaleDateString("ja-JP", { year: "numeric", month: "long", day: "numeric" });
  const msg = `<p>Thank you for your message. I am currently out of office and will return on ${returnStr}.</p>
<p>I will respond to your email as soon as possible upon my return.</p>
<p>If you need immediate assistance, please contact my colleague.</p>
<hr/>
<p>メールありがとうございます。現在不在にしており、${returnStrJa}に戻る予定です。</p>
<p>返信までしばらくお待ちください。</p>`;
  return signature ? `${msg}${signature}` : msg;
}

export function buildOofMessages(endDate: Date, signature: string) {
  return {
    internal: buildInternalMessage(endDate, signature),
    external: buildExternalMessage(endDate, signature),
  };
}

function toISODateTimeStr(date: Date): string {
  return `${toISODateStr(date)}T00:00:00`;
}

export async function setOof(settings: OofSettings): Promise<void> {
  const token = await getAccessToken();
  const scheduledEndDate = addDays(settings.endDate, 1);

  const body = {
    automaticRepliesSetting: {
      status: "scheduled",
      scheduledStartDateTime: {
        dateTime: toISODateTimeStr(settings.startDate),
        timeZone: "UTC",
      },
      scheduledEndDateTime: {
        dateTime: toISODateTimeStr(scheduledEndDate),
        timeZone: "UTC",
      },
      internalReplyMessage: settings.internalMessage,
      externalReplyMessage: settings.externalMessage,
      externalAudience: "all",
    },
  };

  const resp = await fetch("https://graph.microsoft.com/v1.0/me/mailboxSettings", {
    method: "PATCH",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  if (!resp.ok) {
    const err = await resp.text();
    throw new Error(`Failed to set OOF: ${resp.status} ${err}`);
  }
}

File: OutOfOfficeAddin/Services/OofService.cs
using System;
using System.Globalization;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading.Tasks;
using Newtonsoft.Json;

namespace OutOfOfficeAddin.Services
{
    /// <summary>
    /// Sets the user's Outlook automatic-reply (OOF) settings via Microsoft Graph.
    /// </summary>
    public class OofService
    {
        private readonly GraphAuthService _auth;

        public OofService(GraphAuthService auth)
        {
            _auth = auth;
        }

        /// <summary>
        /// Enables automatic replies for the given date range.
        /// </summary>
        /// <param name="startDate">First day of absence.</param>
        /// <param name="endDate">Last day of absence (back-date text will show endDate + 1).</param>
        /// <param name="signatureHtml">User's HTML signature to append to both messages.</param>
        public async Task SetAsync(DateTime startDate, DateTime endDate, string signatureHtml)
        {
            var token = await _auth.AcquireTokenAsync();
            var backDate = endDate.AddDays(1);

            var internalHtml = BuildInternalHtml(backDate, signatureHtml);
            var externalHtml = BuildExternalHtml(backDate, signatureHtml);

            var payload = new
            {
                automaticRepliesSetting = new
                {
                    status = "scheduled",
                    scheduledStartDateTime = new
                    {
                        dateTime = startDate.Date.ToString("yyyy-MM-ddTHH:mm:ss", CultureInfo.InvariantCulture),
                        timeZone = "Tokyo Standard Time"
                    },
                    scheduledEndDateTime = new
                    {
                        // End at 23:59 on the last day so the OOF covers the full day
                        dateTime = endDate.Date.AddDays(1).AddSeconds(-1)
                            .ToString("yyyy-MM-ddTHH:mm:ss", CultureInfo.InvariantCulture),
                        timeZone = "Tokyo Standard Time"
                    },
                    internalReplyMessage = internalHtml,
                    externalReplyMessage = externalHtml
                }
            };

            var json = JsonConvert.SerializeObject(payload);
            using (var client = new HttpClient())
            {
                client.DefaultRequestHeaders.Authorization =
                    new AuthenticationHeaderValue("Bearer", token);

                var content = new StringContent(json, Encoding.UTF8, "application/json");

                // PatchAsync is .NET 5+; use SendAsync with HttpMethod.Patch for .NET Framework 4.8
                var request = new HttpRequestMessage(new HttpMethod("PATCH"),
                    "https://graph.microsoft.com/v1.0/me/mailboxSettings")
                {
                    Content = content
                };

                var response = await client.SendAsync(request);

                if (!response.IsSuccessStatusCode)
                {
                    var body = await response.Content.ReadAsStringAsync();
                    throw new InvalidOperationException(
                        $"Graph API returned {(int)response.StatusCode}: {body}");
                }
            }
        }

        // ------------------------------------------------------------------ helpers

        private static string BuildInternalHtml(DateTime backDate, string signatureHtml)
        {
            // English date: MMM dd, yyyy (invariant English month abbreviation)
            var engDate = backDate.ToString("MMM dd, yyyy", CultureInfo.GetCultureInfo("en-US"));

            var body =
                $"<p>Dear Sender,</p>" +
                $"<p>Thank you for your email.<br/>" +
                $"I will be back {engDate}. Email will be read with delay.</p>";

            return WrapHtml(body, signatureHtml);
        }

        private static string BuildExternalHtml(DateTime backDate, string signatureHtml)
        {
            var engDate = backDate.ToString("MMM dd, yyyy", CultureInfo.GetCultureInfo("en-US"));
            var jpDate = backDate.ToString("yyyy/MM/dd", CultureInfo.InvariantCulture);

            var body =
                $"<p>Dear Sender,</p>" +
                $"<p>Thank you for your email.<br/>" +
                $"I will be back {engDate}. Email will be read with delay.</p>" +
                $"<p>ご連絡ありがとうございます。<br/>" +
                $"申し訳ありませんが、{jpDate} まで不在のため対応できません。<br/>" +
                $"ご理解いただけますと幸いです。</p>";

            return WrapHtml(body, signatureHtml);
        }

        private static string WrapHtml(string body, string signatureHtml)
        {
            var sig = string.IsNullOrWhiteSpace(signatureHtml)
                ? string.Empty
                : $"<hr/>{signatureHtml}";

            return $"<html><body>{body}{sig}</body></html>";
        }
    }
}
